jwplatform/upload.py

```diff
@@ -0,0 +1,137 @@
+
+        # Mark upload as complete
+        self._mark_upload_completion()
+
+    def _upload_part(self, bytes_chunk, part_number, upload_links):
+        # Add a S3 server-side checksum validation too if possible.
+        computed_hash = self._compute_part_hash(bytes_chunk)
+
+        # Check if the file has already been uploaded and the hash matches. Return immediately without doing anything
+        # if the hash matches.
+        upload_hash = upload_links[part_number - 1]["etag"] if "etag" in upload_links[part_number - 1] else None
+        if upload_hash:
+            if repr(upload_hash) == repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+                return
+
+        upload_link = upload_links[part_number - 1]["upload_link"] if "upload_link" in upload_links[part_number - 1] \
+            else None
+        if not upload_link:
+            raise Exception(f"Invalid upload link for part {part_number}.")
+
+        resp = requests.put(upload_links[part_number - 1]["upload_link"], data=bytes_chunk)
+        resp.raise_for_status()
+
+        returned_hash = resp.headers['ETag']
+        if repr(returned_hash) != repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+            raise DataIntegrityError("The hash of the uploaded file does not match with the hash on the server.")
+        self.logger.info(f"Successfully uploaded part {part_number} for upload id {self.upload_id}")
+
+    def _get_pre_signed_part_links(self, part_count) -> {}:
+        query_params = {'page_length': part_count}
+
+        resp = self.client.list(resource_name='uploads', resource_id=self.upload_id, subresource_name='parts',
+                                query_params=query_params)
+
+        body = resp.json_body
+        # Process the results if there are multiple pages
+        total_page_count = body['total'] // part_count
+        parts = body['parts']
+        if total_page_count > 1:
+            for page_number in range(1, total_page_count):
+                query_params['page'] = page_number + 1
+                resp = self.client.list(resource_name='uploads', resource_id=self.upload_id, subresource_name='parts',
+                                        query_params=query_params)
+                body = resp.json_body
+                parts.extend(body['parts'])
+        return parts
+
+    def _compute_part_hash(self, bytes_chunk) -> str:
+        hashing_instance = md5()
+        hashing_instance.update(bytes_chunk)
+        return hashing_instance.hexdigest()
+
+    def _mark_upload_completion(self):
+        self.client.complete(self.upload_id)
+        self.logger.info("Upload successful!")
+
+
+class SingleUpload:
+
+    def __init__(self, upload_link, file, retry_count):
+        self.upload_link = upload_link
+        self.upload_retry_count = retry_count
+        self.file = file
+
+    def upload(self):
+        # Upload to S3 directly
```

Comment:
Probably not worth calling this out since:
* MultipartUpload also uploads to S3 directly.
* It's not necessarily true that uploads go to S3. In practice this is true today but we could make changes without breaking this client.